seccomp/tracer.go

```diff
@@ -171,59 +189,62 @@ func (t *syscallTracer) start() error {
 	return nil
 }
 
-// Method keeps track of all the 'tracee' processes served by a syscall tracer.
-// From a functional standpoint, this routine acts as a garbage-collector for
-// this class. Note that no concurrency management is needed here as this
-// method runs within its own execution context.
-func (t *syscallTracer) sessionsMonitor() error {
+func (t *syscallTracer) addSession(s seccompSession) error {
 
-	// seccompSession DB to store 'tracees' relevant information.
-	var seccompSessionMap = make(map[uint32]seccompSession)
+	t.seccompSessionMu.Lock()
+	t.seccompSessionMap[s.pid] = s
+	t.seccompSessionMu.Unlock()
 
-	// Launch pidmonitor task at 100ms sampling rate.
-	pm, err := pidmonitor.New(&pidmonitor.Cfg{100})
-	if err != nil {
-		logrus.Error("Could not initialize pidMonitor logic")
-		return err
+	err := t.pm.AddEvent([]pidmonitor.PidEvent{
+		pidmonitor.PidEvent{
+			Pid:   s.pid,
+			Event: pidmonitor.Exit,
+			Err:   nil,
+		},
+	})
+
+	logrus.Debugf("Added session for seccomp-tracee: %v", s)
+	return err
+}
+
+func (t *syscallTracer) removeSession(pid uint32) {
+
+	t.seccompSessionMu.Lock()
+
+	s, ok := t.seccompSessionMap[pid]
+	if !ok {
+		logrus.Errorf("Unexpected error: seccomp fd not found for pid %d", pid)
+		t.seccompSessionMu.Unlock()
+		return
+	}
+
+	if err := syscall.Close(int(s.fd)); err != nil {
+		logrus.Fatal(err)
```

Comment:
I know that i was the one that added this 'fatal' instruction here, but i'm specially concerned given that you are considerably expanding the pidMonitor interval, so there's more room now for Close() to return error if kernel (and other goroutines) has fully cleaned all process' resources. My point: can you substitute this 'fatal' log by a 'warning'?

Reply:
Actually, now that i think about it, this Close() syscall shouldn't be needed at all as the process should be dead by now ...

Reply:
I disagree. This is closing an fd owned by sysbox-fs, even though the fd tracks seccomp on a process. Even if the traced process dies, the fd still needs to be closed (otherwise sysbox-fs will have an file-open leak). And closing the fd should not return an error, because it's simply closing a sysbox-fs owned fd; thus, I think it's fine to log a strong error when this occurs.

Am I missing something here?

Reply:
You are right that sysbox-fs must close their side of the seccomp pipe (fd). But I had some doubts about this  'logrus.Fatal' instruction here as it would bring the entire sysbox-fs down (panic). On the other hand, I now realize that this method (removeSession) runs within the scope of sessionsMonitor common thread, so we wouldn't be reducing the failure-scope in any way by removing this 'panic'.

Reply:
I went ahead and modified the logrus.Fatal() -> logrus.Errorf(), as the former will stop the sysbox-fs process.